### PR TITLE
feat: Add text! macro

### PR DESCRIPTION
**Pull request by kdheepak**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -145,6 +145,19 @@ let line = line!["hello", format!("{name}")];
 let line = line!["bye"; 2];
 ```
 
+## Line
```
Comment (EdJoPaTo) [inline README.md:148]:
```suggestion
## Text
```

File: README.md [hunk 2]
```diff
@@ -145,6 +145,19 @@ let line = line!["hello", format!("{name}")];
...
+
+The `text!` macro creates a `Text` that contains a sequence of lines. It is similar to the `vec!` macro.
+
+```rust
+use ratatui::prelude::*;
+use ratatui_macros::{span, line, text};
+
+let name = "world!";
+let text = text![span!(Modifier::BOLD; "hello"), line![format!("{name}")]];
```
Comment (EdJoPaTo) [inline README.md:157]: First the simple thing, line in a text. Then the more interesting thing, a span can also be a line.
```suggestion
let text = text![line![format!("{name}")], span!(Modifier::BOLD; "hello")];
```

Also… `"hello".bold()` is shorter than the span macro call. The only benefit would be when `span!()` works in const contexts?

File: src/text.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+/// [`Text`]: crate::text::Text
+/// [`Line`]: crate::text::Line
+/// [`Span`]: crate::text::Span
+#[macro_export]
+macro_rules! text {
+    () => {
+        ratatui::text::Text::default()
+    };
+    ($line:expr; $n:expr) => {
+      ratatui::text::Text::from(vec![$line.into(); $n])
```
Comment (EdJoPaTo) [inline src/text.rs:44]: same as https://github.com/ratatui-org/ratatui-macros/pull/29#pullrequestreview-2041957891

`cargo fmt` with some macro specific settings (requires nightly) is likely a good idea in this repo (and CI)

File: src/text.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    use ratatui::prelude::*;
+
+    #[test]
+    fn text() {
+        // literal
+        let text = text!["hello", "world"];
+        assert_eq!(text, Text::from(vec!["hello".into(), "world".into()]));
+
+        // explicit use of span and line
+        let text = text![crate::span!("hello"), crate::line!["world"]];
```
Comment (EdJoPaTo) [inline src/text.rs:66]: same as in doc comment

```suggestion
        let text = text![crate::line!["world"], crate::span!("hello")];
```

**Review by EdJoPaTo — DISMISSED** [review]

File: README.md (hunk 2 above)
Comment (kdheepak) [reply README.md:157]: I wanted to show an example of using `span!` but I'll also add the `"hello".bold()` on the next line so users know they can do that.

File: src/text.rs (hunk 3 above)
Comment (kdheepak) [reply src/text.rs:44]: I did run it and it doesn't change any of the lines in this PR

image

It does make a change that I don't like for the `Fill` macro but I could live with that if we added it to CI.

Comment (EdJoPaTo) [reply src/text.rs:44]: Strange. The indentation is different for this line which seems to be not formatted then.

**Review by joshka — APPROVED** [review]

**Pull request merged**